### PR TITLE
Add support for buffer donation in `jit` and `pmap`.

### PR DESCRIPTION
**Pull request by tomhennigan**
For a computation of the form:

    >>> f = lambda x: x ** 2
    >>> f = jax.jit(f)
    >>> while run:
    ...   x = f(x)

JAX must currently always have two copies of `x` in device memory since there
is no reliable way in Python to determine whether there will be future uses of
`x`. This causes two classes of problem:

  1. Users at the limit of available device are constrained by the additional
     copy of their parameters and other state while they typically only require
     one copy. This typically frees 100M+ of device memory and is a critical
     optimization for larger models to match state of the art performance in
     other frameworks.

  2. This constant alloc/free of the input/output buffers can cause memory
     fragmentation on some platforms (although having a reusing allocator and
     limiting run-ahead may be a better solution for this problem).

We propose fixing this by using input/output aliasing as supported by XLA. We
will support this in JAX by allowing certain arguments of jit/pmap decorated
functions to be donated and reused as outputs:

    >>> f = lambda x: x ** 2
    >>> f = jit(f, donate_argnums=0)
    >>> while run:
    ...   x = f(x)

JAX will determine that the donated input `x` can alias with the output of the
function and it will instruct XLA it _must_ write the result to this buffer.

If a user tries to reuse a buffer after it has been donated they get an error
that the buffer is invalid:

    >>> y = f(x)
    >>> jax.device_get(x)
    ...
    RuntimeError: Invalid argument: CopyToHostAsync() called on invalid buffer.

The semantics of `donate_argnums` follows that of `static_argnums`, namely that
it identifies positional arguments to the computation that are to be donated
to the computation and used as part of the output.

One feature that is also enabled by this is invalidating buffers that should
only be used once, for example PRNGKeys:

    >>> @partial(jit, donate_argnums=0)
    ... def move(x):
    ...   # Do something complex enough for JAX to just optimize it away.
    ...   return tree_map(lambda x: x + x - x, x)

    >>> def safe_eager_uniform(key, *a, **k):
    ...   assert hasattr(key, 'device_buffer'), "random must run eagerly"
    ...   key = move(key)
    ...   return jax.random.uniform(key, *a, **k)

This is not a complete answer to random safety since it is still possible to
reuse a key as part of a traced computation, however it can be used to support
this feature (somewhat inefficiently) in eager mode.

### CODE REVIEW

**Comment by tomhennigan** [comment]
@hawkinsp @skye FYI

File: jax/api_util.py
Comment (gnecula) [inline jax/api_util.py:107]: This assumes that static_argnums are non-repeating.

File: jax/interpreters/pxla.py
Comment (gnecula) [inline jax/interpreters/pxla.py:569]: This code feels dangerous, we assume that whoever added args, did so only at the end. Wouldn't it be easier to update donate whereever args are added, so that by the time we get here we know that donate corresponds to args. I think that the logic should be very similar to updating mapped_invars (which was tricky).

File: jax/api.py
Comment (gnecula) [inline jax/api.py:1086]: Would it make sense to call this donate_invars (following mapped_invars)?

File: jax/interpreters/xla.py
Comment (gnecula) [inline jax/interpreters/xla.py:470]: xe.Device failed mypy for me. I have been using just 'Device' defined at top of the file.

File: jax/interpreters/xla.py
Comment (gnecula) [inline jax/interpreters/xla.py:617]: Maybe a type hint or comment would help?

File: jax/interpreters/xla.py
Comment (gnecula) [inline jax/interpreters/xla.py:748]: The exception message suggests that the check should be `if any(donate)`.

File: jax/interpreters/xla.py
Comment (gnecula) [inline jax/interpreters/xla.py:760]: This is suspicious, why should we let `donate` get out of sync with args?

File: tests/api_util_test.py
Comment (gnecula) [inline tests/api_util_test.py:58]: Thanks for the unit tests!

File: jax/interpreters/xla.py
Comment (gnecula) [inline jax/interpreters/xla.py:618]: This code is complex enough that it deserves unit tests?

File: tests/api_test.py
Comment (gnecula) [inline tests/api_test.py:3183]: I am curious where is the logic that deletes the device buffer implemented?

**Review by gnecula — COMMENTED** [review]
You definitely need to get mattjj@ to look at this. When I have added `mapped_invars` I missed several places where it should be updated (https://github.com/google/jax/pull/2828)

File: jax/api.py
Comment (gnecula) [inline jax/api.py:121]: Should we document that this works only on TPUs? What happens on other platforms. I assume there is no error, just no aliasing.

Comment (tomhennigan) [reply jax/api.py:121]: There is an error.

File: tests/api_test.py
Comment (tomhennigan) [reply tests/api_test.py:3183]: This happens in the XLA runtime when you pass a buffer to a computation that has input/output aliasing.

File: jax/interpreters/pxla.py
Comment (tomhennigan) [reply jax/interpreters/pxla.py:569]: Ah thanks for noticing this, we should follow the logic in xla.py for donate and raise an error `if any(donate)`. AFAIk this happens only for `jvp(jit(.., donate_argnums=..))`.

File: jax/api.py
Comment (tomhennigan) [reply jax/api.py:1086]: I feel like semantically it is very close to static_argnums (e.g. it identifies arguments by index that are donated). Is there a reason to prefer invars over argnums?

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py:470]: Will address all these type issues once the comments are in from other folks. FYI the tests all pass internally (with pytype checking the types).

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py:760]: I was thinking that transpose of `f(x, y) -> z` is `f'(z) -> dx, dy`? Perhaps I've totally misunderstood?

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py:748]: 🤦 yep that would be a lot simpler 😄

File: jax/api.py
Comment (gnecula) [reply jax/api.py:1086]: My thinking was that unlike static_argnums, donate is carried through the Jaxpr and it describes the behavior of the invars.

File: jax/api.py
Comment (trevorcai) [reply jax/api.py:121]: I personally preferred this without the error, as it lets me run the same code for both local GPU & TPU cases. I now need to check devices & grab their platform to decide my pmap args!
(and I already expect that memory allocation behavior is different between GPU & TPU, so am not surprised that I can't run to the same efficiency on GPU.)

File: jax/interpreters/xla.py
Comment (skye) [inline jax/interpreters/xla.py:635]: High-level question: why look for matching outputs instead of having the user specify explicit (input, output) donation pairs? I'm guessing it's more ergonomic to only specify the donated inputs, and maybe more convenient if you don't have specific outputs in mind. The downsides I see are that this could theoretically lead to the wrong matching if there are multiple inputs/outputs with the same shape, and the API is a little subtle since you get an error if there are unmatched inputs (although including the output requirements in the docstrings would probably be sufficient).

(Apologies if I missed some earlier discussion somewhere...)

File: jax/api.py
Comment (skye) [reply jax/api.py:121]: +1, I'm not a real user so listen to Trevor more than me, but I like the idea of not needing to modify code to switch platforms (modulo low-level details like memory management...). Maybe issue a one-time warning if used on GPU?

**Review by skye — COMMENTED** [review]
This is really nice! I didn't do a super in-depth review, please let me know if you want one from me (and sorry for taking a few days to look at this, I'll be more prompt with future review).

File: jax/api_util.py
Comment (hawkinsp) [inline jax/api_util.py:97]: It wasn't immediately clear to me what "rebase" meant (I figured it out after a few seconds of thinking), and I think an example would make it clear right away.

File: jax/interpreters/pxla.py
Comment (hawkinsp) [inline jax/interpreters/pxla.py:846]: Every time you use an unnecessary newline character @mattjj dies a little inside...

(This is fine, I jest... but we should all pitch in and buy him a larger monitor.)

File: jax/interpreters/xla.py
Comment (hawkinsp) [reply jax/interpreters/xla.py:470]: I need to add a `.pyi` file for the XLA extension I think. I'll get on it soonish.

File: jax/interpreters/xla.py
Comment (hawkinsp) [inline jax/interpreters/xla.py:480]: Bad line wrapping.

File: jax/interpreters/xla.py
Comment (hawkinsp) [inline jax/interpreters/xla.py:475]: Perhaps @mattjj has input here? This feels a bit strange to me.

File: jax/interpreters/xla.py
Comment (hawkinsp) [inline jax/interpreters/xla.py:493]: Claim: JAX only uses flat shapes these days. We use tuples for exactly two reasons:
a) to pass more than N arguments to a computation, where N is a few hundred, and
b) as return values from computations (we may even destructure these before they make it back to JAX.)

Do we still need this function?

File: jax/interpreters/xla.py
Comment (hawkinsp) [reply jax/interpreters/xla.py:635]: I have a similar question. But presumably the donation API could generalize to user-specified aliasing (pass a dict instead of a set).

File: jax/interpreters/xla.py
Comment (hawkinsp) [inline jax/interpreters/xla.py:646]: The phrasing of the API as "donation" makes this somewhat surprising. I'm presumably free to throw your donation away, in the usual meaning of those words?

**Review by hawkinsp — COMMENTED** [review]
Some comments.

File: jax/interpreters/pxla.py
Comment (mattjj) [reply jax/interpreters/pxla.py:569]: Yes, `JVPTrace.process_call` adds arguments to an `xla_call` (and `JVPTrace.process_map` adds arguments to an `xla_pmap`), so we should update the `params` we pass to `call_primitive.bind` (and `map_primitive.bind`, respectively) in those functions. However, we should also keep in mind that `JaxprTrace.process_call` (and `JaxprTrace.process_map`) produces **two** calls, one with additional outputs and the other with additional inputs; those arises in grad-of-jit situations, for example.

In particular, in `JVPTrace.process_call` we should extend [`donate` in these `params`](https://github.com/google/jax/blob/0ea22b7e191a8cfabefeae18834ced2eec80670b/jax/interpreters/ad.py#L344) by the length of `nonzero_tangents`. (It would be safe to always mark them as False, though there could be another policy, like the tangents corresponding to donated primals are always themselves donated.) Something unfortunate is that we have several call primitives and that same `process_call` is used for all of them, yet only `xla_call` has the `donate` parameter. As a result, we won't always have a `"donate"` in `params`. I'm sure we can figure out good ways to organize things, though I'm fine with anything (e.g. just checking if `"donate"` is in `params` and updating it if it's there).

In `JVPTrace.process_map`, we should also extend [`donate` in these `new_params`](https://github.com/google/jax/blob/0ea22b7e191a8cfabefeae18834ced2eec80670b/jax/interpreters/ad.py#L373). There `mapped_invars` / `new_mapped_invars` provides a good example of the kind of thing we want to do.

I'll write a follow-up comment about `JaxprTrace.process_call` and `JaxprTrace.process_map`.

File: jax/interpreters/pxla.py
Comment (mattjj) [reply jax/interpreters/pxla.py:846]: Guys I just have my 13" laptop at home, I'm dying.

File: tests/api_test.py
Comment (skye) [reply tests/api_test.py:3183]: Forgot to mention earlier, the upshot of marking deleted in Python is we can provide better error messages (instead of relying on errors from the runtime). But I think runtime errors are fine for now.

File: jax/interpreters/xla.py
Comment (skye) [reply jax/interpreters/xla.py:475]: Are you worried about the grad-of-jit case? I think it should error before it hits this. Maybe add a test case to make sure it errors nicely?

File: jax/interpreters/pxla.py
Comment (skye) [reply jax/interpreters/pxla.py:846]: Honestly I love this PR, EXCEPT for these super tall argument lists :P

File: tests/api_test.py
Comment (trevorcai) [reply tests/api_test.py:3183]: I wouldn't block this PR on this, but I think it's important to give an error message couched in terms of JAX terminology (e.g. donation).

Anecdote: I tripped up into reusing a donated argument in a test a week after I patched this. It took me a while to realize that the errors I had were around reusing a donated buffer cuz the errors generated by the runtime are pretty oblique.
I signed up for this by being an early adopter, but ideally future users shouldn't also need to interpret TPU runtime error messages.

**Comment by avital** [comment]
This looks very exciting! Is there any plan to add support for GPU as well?

**Comment by mattjj** [comment]
@avital my understanding is that this is platform-agnostic from the JAX end, but we'd need to request support for this feature from the XLA:GPU folks.

**Comment by tomhennigan** [comment]
I asked a few people to meet with me to review the design of this PR. Overall the proposed API was approved by the group and we will now iterate on a few changes to the implementation that will not be user visible (e.g. interaction with JVPTrace).

One major change as a result of the review is that we are leaning in to "gifting" nature of "donation" (previously we would error if we could not use a donation, now we will simply discard it). The original implementation focused exclusively on the case where a donated buffer is aliased with an output, however there are other future cases where donation may make sense (e.g. reusing an input buffer for an intermediate computation) and on some backends we may not be able to make use of donated buffers (e.g. on CPU/GPU input/output aliasing is not supported).

We decided that donation will be best effort and on backends that don't offer anything to do with donated buffers (e.g. CPU/GPU) we will log a warning and invalidate the the Python objects (to avoid accidental reuse). We will additionally no longer require perfect aliasing of inputs and outputs (meaning that on backends which do support donation you can donate more than you need).

There were concerns raised about providing the user control over which specific buffers in the input aliased the output. For now we will document that input/output aliasing is managed by traversing the flat input and flat output in order and matching the first output with an appropriate shape/dtype to the input. This means you can have fine grained control over aliasing through re-ordering outputs in Python.

Thanks to everyone who participated for an engaging discussion! I hope to have the patch ready to submit by the end of the month.

**Comment by mattjj** [comment]
Heads up, might need a rebase on #3210 when that goes in. Since that PR aims to be a cleanup, hopefully if there's any conflict it's only to make this PR easier.

File: jax/api_util.py
Comment (tomhennigan) [reply jax/api_util.py]: Fixed.

File: jax/interpreters/pxla.py
Comment (tomhennigan) [reply jax/interpreters/pxla.py]: Have fixed this properly with @mattjj's help, we now correctly handle this following mapped_invars (extending to call as well as map).

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py]: I've dropped these hints for now based on other comments re arg formatting.

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py]: See above comments wrt JVPTrace, we are now handling this correctly where inputs to call primitives are changed and I've removed this.

File: tests/api_test.py
Comment (tomhennigan) [reply tests/api_test.py:3137]: When a buffer is donated and used as part of a computation it is marked as deleted and further uses will raise an error. We discussed leaving a tombstone on the buffer object to give better errors, I'll leave this for a follow up.

File: jax/api.py
Comment (tomhennigan) [reply jax/api.py:117]: Per the discussion in the design review I've made consumption of donated buffers optional and as of this PR all we'll do is setup "must aliases" on TPU iff we find a match between the input and the output buffers.

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py]: As per discussion in design review we'll leave this API as is for now (just specifying what buffers you are letting JAX reason about and not telling us specifically what to do with them).

File: jax/api_util.py
Comment (tomhennigan) [reply jax/api_util.py]: Have added a comment.

File: jax/interpreters/pxla.py
Comment (tomhennigan) [reply jax/interpreters/pxla.py]: Heh I have a different opinion 😄 . Have reverted this (and all other signatures) to the current "condensed and lightly typed" style matching the rest of the file.

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py]: With @mattjj's help this is now resolved. properly 😄

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py:535]: I think its good to have this function to handle that case, my preference would be to keep the general implementation since it is not all that complex. I guess we could shave a few lines off if we made this switch on tuple_args. Please feel free to reopen if you feel strongly.

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py]: Per the design review I think we're good with "donation" being "use if you like" (allowing us to not use them at all on CPU/GPU today and still be "in contract")/

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py:650]: We test the result of this (e.g. on supported platforms input buffers are deleted and output, if there is no match we raise a warning) but I haven't unit tested the method directly, it feels like an implementation detail to me. Do you feel strongly that this is tested in isolation? To do this well I would need to refactor it a bit to not require us to build up a computation.

Comment (tomhennigan) [reply jax/interpreters/xla.py:650]: I'm going to assume the current tests are fine and close this out, but please re-open if you feel strongly.

File: jax/api.py
Comment (tomhennigan) [reply jax/api.py]: Have kept this as `donate_argnums` on the API and `donate_invars` on the primitive.

File: jax/interpreters/xla.py
Comment (tomhennigan) [reply jax/interpreters/xla.py:649]: I think this method is pretty clear now, adding types would probably complicate the code a fair bit and not add too much value.

**Comment by tomhennigan** [comment]
Thanks for the comments everyone, **this PR is ready for your review again** 😄  I've rebased on master and made some major changes based on the design review:

1. We now interact correctly with JVPTrace, following the guidance of `mapped_invars` (thanks for the help @mattjj!!).
2. Donation is now optional, on CPU/GPU we simply log a warning at "compile" time that the buffers aren't going to be used and do nothing else.
3. Function arguments and type hints are now consistent with the rest of JAX.

PTAL!

File: jax/api.py
Comment (mattjj) [inline jax/api.py:118]: Maybe instead of "XLA will use ..." we can say "XLA can reuse their memory, e.g. to store the result of the computation."

However, I'm fine with merging first and making docstring tweaks in follow-ups, so this isn't blocking.

File: jax/api.py
Comment (mattjj) [inline jax/api.py:984]: Same docstring suggestion here.

**Comment by tomhennigan** [comment]
FYI two known issues that I think we can address in a follow up:

1. If you try and donate the same array twice in the same computation (e.g. `jit(f, donate_argnums=(0, 1))(x, x)`) you get a deadlock.
2. Buffers that are donated but not used are currently not marked as "deleted" meaning that you can reuse them (this is not a correctness issue, but might lead to confusing "this buffer is deleted" errors if you move your code to TPU where input/output aliasing is supported). In a follow up we should delete these buffers at call time.

File: jax/core.py
Comment (mattjj) [inline jax/core.py:1206]: I think this check should be hoisted one level up, i.e. it shouldn't only apply to the `if prim.map_primitive` case but rather to the `if prim.call_primitive or prim.map_primitive` case.

File: jax/interpreters/pxla.py
Comment (mattjj) [inline jax/interpreters/pxla.py:732]: Can we rename `setup_aliases` to `set_up_aliases`? After all, "setup" is a noun and "set up" is a verb phrase, and the latter is intended.

**Review by mattjj — APPROVED** [review]
LGTM modulo the few suggestions offered, mostly docs/names.

File: jax/core.py
Comment (tomhennigan) [reply jax/core.py:1206]: Happy to do so, but we'd need to match on primitive name which doesn't feel ideal, since donation doesn't apply to all `call` primitives (e.g. remat does not have this argument) but it does apply to all (well, the only) `map` primitive.

File: jax/interpreters/pxla.py
Comment (tomhennigan) [reply jax/interpreters/pxla.py]: Done. I was following `c.setup_alias`, I think thats a typo too so I'll fix it in a follow up.

File: jax/api.py
Comment (tomhennigan) [reply jax/api.py]: Ah thanks, I hadn't updated the docstring to match the implementation. Have reworded.

File: jax/api_util.py
Comment (jekbradbury) [inline jax/api_util.py:118]: This is gonna be super slow, but it's only invoked at compile time so it doesn't really matter.

File: jax/core.py
Comment (jekbradbury) [reply jax/core.py:1206]: We might want `if prim.consuming_primitive` or something like that, but no need to do that now.

File: jax/interpreters/ad.py
Comment (jekbradbury) [inline jax/interpreters/ad.py:486]: I have a vague sense that we might just be able to make this `True` (assuming that `backward_pass` is only called inside the AD infra) because it only has an effect in the op-by-op grad-of-jit case, and in that case using donation for cotangent buffer accumulation is always a good idea. But let's follow up and not make that change now!

File: jax/core.py
Comment (mattjj) [reply jax/core.py:1206]: Oh yeah, good point, I forgot! Nevermind.

File: jax/interpreters/partial_eval.py
Comment (jekbradbury) [inline jax/interpreters/partial_eval.py:192]: Similarly, in the case where `partial_eval` is used for reverse-mode AD (but only `grad`, not `linearize`,  because that's all about reusing residuals!), it's possible that we could use donation on `const_tracers` because we know they'll never be reused.

**Review by jekbradbury — APPROVED** [review]

File: jax/interpreters/ad.py
Comment (mattjj) [reply jax/interpreters/ad.py:486]: Hey nice catch, I think you're right!

File: jax/interpreters/partial_eval.py
Comment (mattjj) [reply jax/interpreters/partial_eval.py:192]: Indeed as you say this wouldn't work for `linearize` (of `jit`), but it also wouldn't work for `vjp`, because in both cases we might call the linearized function many times and hence can't throw away the residuals after the first execution. It'd only work for `grad`, though of course `grad` is an important special case and might be worth thinking about lightweight ways to provide that optimization.

**Comment by jekbradbury** [comment]
Filed https://github.com/google/jax/issues/3286 to keep track of AD-related buffer donation opportunities.

**Pull request merged**
